File: src/main/scala/scala99/P09.scala
package scala99

/**
  * Created by rick on 2017/3/24.
  */
object P09 {
    def pack[T](list: List[T]): List[List[T]] = {
        if (list == null || list.isEmpty)
            Nil
        else {
            val (before, after) = list.span(x => x == list.head)
            before +: pack(after)
        }
    }
}

File: src/main/kotlin/kotlin99/P09.kt
package kotlin99
/**
 * Created by rick on 2017/2/17.
 */
fun <T> pack(list: List<T>): List<List<T>> {
    if (list.isEmpty())
        return emptyList()
    else {
        val result = list.takeWhile {
            it == list[0]
        }
        return listOf(result) + pack(list.takeLast(list.size - result.size))
    }
}

